README.md
# <img alt="puppeteer screen recorder logo" width="128px" src="https://github.com/lukaselmer/puppeteer-screen-recorder/blob/main/asserts/puppeteer-screen-recorder.png" /> puppeteer-screen-recorder

A puppeteer Plugin that uses the native
[chrome devtool protocol](https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-startScreencast)
for capturing video frame by frame. Also supports an option to follow pages that are opened by the
current page object.
[Check out API Docs](https://lukaselmer.github.io/puppeteer-screen-recorder/classes/puppeteerscreenrecorder.html).

[![NPM](https://nodei.co/npm/puppeteer-screen-recorder.png)](https://npmjs.org/package/puppeteer-screen-recorder)

![GitHub](https://img.shields.io/github/license/lukaselmer/puppeteer-screen-recorder)

![GitHub package.json version](https://img.shields.io/github/package-json/v/lukaselmer/puppeteer-screen-recorder)
![GitHub top language](https://img.shields.io/github/languages/top/lukaselmer/puppeteer-screen-recorder)
![npm](https://img.shields.io/npm/dt/puppeteer-screen-recorder)
![Snyk Vulnerabilities for npm package](https://img.shields.io/snyk/vulnerabilities/npm/puppeteer-screen-recorder-improved)

<a href="https://github.com/lukaselmer/puppeteer-screen-recorder/issues/new">Add Feature Request</a>

</p>

## Be a Sponsor

Puppeteer-screen-recorder isn't backed by a company, so the future of this project depends on you. Become
a sponsor or a backer - help the open source community.

#### **For Companies and Businesses**

By becoming a Github Sponsor, your company and brand will be recognized as a one that gives back to the
open source tools that run your business and one that respects your developers time and your customers'
experience.

#### **For Developers**

By helping your company become a Github Sponsor, you will not only feel great about giving back to the
open source tools that run your business run, If you believe your company could become a sponsor, then
please reach out!

- See https://github.com/prasanaworld/puppeteer-screen-recorder/

## Key Feature

### 1. Follow Page Automatically

Automatically follows pages (multiple pages) which are opened at runtime, which will be part of video
capturing. Also support options to disable the default flow.

### 2. No overhead over FF_MPEG library

FFMPEG library's installation and configuration are automatically managed by the library internally. Also
offers options to configure with custom library path.

### 3. Native Implementation

This plugin works directly with native
[chrome devtool protocol](https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-startScreencast)
to capture the video under the wood without any other thirdparty puppeteer plugins for screen capturing.

### 4. Adopted the Chromium principles

Adopted Chromium principles such as Speed, Security, Stability and Simplicity. It also ensures no frames
are missed during video capturing and doesn't impact the performance, since its doesn't use any other
puppeteer plugin internally.

### 5. Supports multiple video format and stream

Supports multiple video format like AVI, MP4, MOV and WEBM. Enable support for writable or duplex stream
for process the output streaming .

## Getting Started

### Installation Guide

Using Npm

```sh
npm install puppeteer-screen-recorder
```

Using Yarn

```sh
yarn add puppeteer-screen-recorder
```

### API Description

**1. Import the plugin using ES6 or commonjs.**

```javascript
// ES6
import { PuppeteerScreenRecorder } from 'puppeteer-screen-recorder'

// or commonjs
const { PuppeteerScreenRecorder } = require('puppeteer-screen-recorder')
```

**2. Setup the Configuration object.**

```javascript
const Config = {
  followNewTab: true,
  fps: 25,
  ffmpegPath: '<path of ffmpeg_path>' || null,
  videoFrame: {
    width: 1024,
    height: 768,
  },
  videoCrf: 18,
  videoCodec: 'libx264', // or libx264, libvpx-vp9
  videoPreset: 'ultrafast',
  videoBitrate: 1000,
  autoPad: {
    color: 'black' | '#35A5FF',
  },
  aspectRatio: '4:3',
}
```

> - **followNewTab** : Boolean value which is indicate whether to follow the tab or not. Default value is
>   true.

> - **fps**: Numeric value which denotes no.of Frames per second in which the video should be recorded.
>   default value is 25.

> - **ffmpegPath**: String value pointing to the installation of [FFMPEG](https://ffmpeg.org/). Default
>   is null (Automatically install the FFMPEG and use it).

> - **videoFrame**: An object which is to specify the width and height of the capturing video frame.
>   Default to browser viewport size.

> - **aspectRatio**: Specify the aspect ratio of the video. Default value is `4:3`.

> - **autoPad**: Specify whether autoPad option is used and its color. Default to autoPad deactivation
>   mode.

> - **recordDurationLimit**: Numerical value specify duration (in seconds) to record the video. By
>   default video is recorded till stop method is invoked`. (Note: It's mandatory to invoke Stop() method
>   even if this value is set)

**3. create a new instance of video recording**

```javascript
const recorder = new PuppeteerScreenRecorder(page, Config) // Config is optional

// or

const recorder = new PuppeteerScreenRecorder(page)
```

> - **page**: Puppeteer page object which needs to captured.
> - **config**: Config is an optional object. Default value is
>   `{ followNewTab: true, fps: 25, ffmpegPath: null }`

**4. Start Video capturing**

**Option 1 - Start video capturing and save as file**

```javascript
const savePath = './test/demo.mp4'
await recorder.statWritingToFile(savePath)
```

**Option 2 - Start Video capturing using stream**

```javascript
const pipeStream = new PassThrough()
await recorder.startWritingToStream(pipeStream)
```

> **pass**: Any writeable stream that will be an output for the stream recorder. Video is recorded and
> streamed with .mp4 extension.

> **savePath**: string value indicating the directory on where to save the video. The path must also
> specify the name of the video with extension .mp4 (example - ./test/puppeteer-demo.mp4). Starting from
> v2, support added for extensions mp4, avi, mov and webm.

**5. Stop the video capturing.**

```javascript
await recorder.stop()
```

### Example

```javascript
const puppeteer = require('puppeteer')
const { PuppeteerScreenRecorder } = require('puppeteer-screen-recorder')

;(async () => {
  const browser = await puppeteer.launch()
  const page = await browser.newPage()
  const recorder = new PuppeteerScreenRecorder(page)
  await recorder.statWritingToFile('./report/video/simple.mp4') // supports extension - mp4, avi, webm and mov
  await page.goto('https://example.com')

  await page.goto('https://test.com')
  await recorder.stop()
  await browser.close()
})()
```

## FAQ

Some of the frequently asked question about puppeteer screen recording are

**Q: Does it support Chrome in headless mode?**

---

Yes, it supports Both headless and headFul mode.

It records full length video, frame by frame even when Chrome runs in headless mode.

**Q: Does it support multiple extension?**

---

Yes, Starting from version 2.0, support multiple extension like AVI, MP4, MOV and WEBM.

**Q: Does it use the window object?**

---

No, it's not tied to the window Object.

**Q: Does it follows pages which are opened at runtime**

---

Yes, it automatically follows pages which is created at runtime.

**Q: is there an option to disable video recording for new page created and record video only for the
page object passed**

---

Yes, By setting the `options.followNewTab` to false, it record only video for the passed page object
alone.

**Q: Does it support to record video at 60 fps**

---

Yes, video can be recorded at 60 fps. By setting `options.fps` to 60.

**Q:Does it use the window object?**

---

No, it doesn't use the window object.

**Q: Does it use FFMPEG internally?**

---

Yes, it uses FFMPEG with optimized options to speed up the video recording using stream from chrome
devtool protocol.

**Q: Does it support Webm?**

---

Webm format is supported.

**Q: Is it possible to output the stream for further processing/enhancing?**

---

Yes. By passing writable stream/duplex stream to `startWritingToStream` method.

**Q: Can I limit the time of recording, like to stop after 2 minutes?**

---

By specifying the time to record (in seconds) using `option.recordDurationLimit`

**Q: Does it support audio recording?**

---

No, it doesn't audio recording.

**Q: Does it support GIF format, any plan to support in fur?**

---

No, it wont support GIF. since Gif is considered as a image format.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See
[standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.1.3](https://github.com/lukaselmer/puppeteer-screen-recorder/compare/v3.1.2...v3.1.3) (2023-08-10)

### [3.1.2](https://github.com/lukaselmer/puppeteer-screen-recorder/compare/v3.1.1...v3.1.2) (2023-08-07)

### [3.1.1](https://github.com/lukaselmer/puppeteer-screen-recorder/compare/v3.1.0...v3.1.1) (2023-08-07)

### [3.1.0](https://github.com/lukaselmer/puppeteer-screen-recorder/compare/v3.0.3...v3.1.0) (2023-08-07)

### [3.0.3](https://github.com/lukaselmer/puppeteer-screen-recorder/compare/v3.0.1...v3.0.3) (2023-08-07)

### [3.0.1](https://github.com/lukaselmer/puppeteer-screen-recorder/compare/v3.0.0...v3.0.1) (2023-08-06)

### [3.0.0](https://github.com/lukaselmer/puppeteer-screen-recorder/compare/v2.1.2...v3.0.0) (2023-08-06)

### [2.1.2](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v2.1.1...v2.1.2) (2022-10-05)

### [2.1.1](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v2.1.0...v2.1.1) (2022-09-17)

## [2.1.0](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v2.0.2...v2.1.0) (2022-08-29)

### Features

- add autoPad option on ffmpeg configuration
  ([6f53e21](https://github.com/prasanaworld/puppeteer-screen-recorder/commit/6f53e21cd76c24a3808d8af24a243234d6a09ddc))

### [2.0.4](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v2.0.3...v2.0.4) (2022-08-27)

### [2.0.3](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v2.0.2...v2.0.3) (2022-08-06)

### [2.0.2](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v2.0.0...v2.0.2) (2021-09-26)

### [2.0.1](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.12...v2.0.1) (2021-09-25)

### [1.1.12](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.11...v1.1.12) (2021-09-25)

### [1.1.11](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.10...v1.1.11) (2021-05-17)

### [1.1.10](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.9...v1.1.10) (2021-05-17)

### [1.1.9](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.8...v1.1.9) (2021-03-28)

### [1.1.8](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.7...v1.1.8) (2021-03-28)

### [1.1.7](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.6...v1.1.7) (2021-03-28)

### [1.1.6](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.5...v1.1.6) (2021-03-04)

### [1.1.5](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.3...v1.1.5) (2021-02-26)

### [1.1.4](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.3...v1.1.4) (2021-02-26)

### [1.1.3](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.2...v1.1.3) (2021-02-25)

### [1.1.2](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.1...v1.1.2) (2021-02-25)

### [1.1.1](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.1.0...v1.1.1) (2021-02-24)

## [1.1.0](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.0.4...v1.1.0) (2021-02-23)

### [1.0.4](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.0.3...v1.0.4) (2021-02-23)

### [1.0.3](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.0.2...v1.0.3) (2021-02-22)

### [1.0.2](https://github.com/prasanaworld/puppeteer-screen-recorder/compare/v1.0.1...v1.0.2) (2021-02-22)

### 1.0.1 (2021-02-22)


src/lib/pageVideoStreamWriter.ts
import { EventEmitter } from 'events'
import os from 'os'
import { PassThrough, Writable } from 'stream'
import ffmpeg, { Codec, Codecs, FfmpegCommand, setFfmpegPath } from 'fluent-ffmpeg'
import {
  PageScreenFrame,
  PuppeteerScreenRecorderOptions,
  VideoWriteStatus,
} from './pageVideoStreamTypes'
import { fileFormatFor, SupportedVideoFileFormat } from './supportedFileFormats'

export class PageVideoStreamWriter extends EventEmitter {
  private readonly screenLimit = 10
  private screenCastFrames: PageScreenFrame[] = []
  duration = '00:00:00:00'

  private status: VideoWriteStatus = 'notStarted'
  private options: PuppeteerScreenRecorderOptions

  private videoMediatorStream: PassThrough = new PassThrough()
  private writerPromise: Promise<void> | undefined

  constructor(
    private destination: string | Writable,
    options: PuppeteerScreenRecorderOptions
  ) {
    super()
    this.options = options
  }

  async startStreamWriter(): Promise<void> {
    await this.configureFfmpegPath()

    if (typeof this.destination === 'string') await this.configureVideoFile(this.destination)
    else if (this.destination.writable) await this.configureVideoWritableStream(this.destination)
    else throw new Error('Output should be a path or a writable stream')
  }

  private async configureVideoFile(destinationPath: string) {
    const outputFormat = fileFormatFor(destinationPath) || this.outputFormat
    const outputStream = await this.getDestinationStream(
      outputFormat,
      this.videoFileCodec(destinationPath)
    )

    this.writerPromise = new Promise((resolve, reject) => {
      outputStream
        .on('error', (e: Error) => {
          this.handleWriteStreamError(e.message)
          reject(e)
        })
        .on('end', resolve)
        .save(destinationPath)

      if (outputFormat == 'webm') {
        outputStream
          .videoBitrate(this.options.videoBitrate || 1000, true)
          .outputOptions('-flags', '+global_header', '-psnr')
      }
    })
  }

  private async configureVideoWritableStream(writableStream: Writable) {
    const outputStream = await this.getDestinationStream(this.outputFormat, this.videoStreamCodec())

    this.writerPromise = new Promise<void>((resolve, reject) => {
      try {
        outputStream
          .on('error', (e) => {
            writableStream.emit('error', e)
            reject(e)
          })
          .on('end', () => {
            writableStream.end()
            resolve()
          })

        outputStream.toFormat(this.outputFormat)
        if (this.outputFormat == 'webm') {
          outputStream
            .videoBitrate(this.options.videoBitrate || 1000, true)
            .addOutputOptions('-flags', '+global_header', '-psnr')
        }
        outputStream.addOutputOptions(
          '-movflags +frag_keyframe+separate_moof+omit_tfhd_offset+empty_moov'
        )

        outputStream.pipe(writableStream)
      } catch (error) {
        reject(error)
      }
    })
  }

  private async getDestinationStream(
    outputFormat: SupportedVideoFileFormat,
    videoCodec: string
  ): Promise<FfmpegCommand> {
    await checkVideoCodec(videoCodec)
    const threads = Math.max(1, os.cpus().length - 1)

    const outputStream = ffmpeg({
      source: this.videoMediatorStream,
      priority: 20,
    })
      .videoCodec(videoCodec)
      .size(this.videoFrameSize)
      .aspect(this.options.aspectRatio || '4:3')
      .autopad(this.autopad.activation, this.autopad?.color)
      .inputFormat('image2pipe')
      .inputFPS(this.fps)
      .outputOptions(`-crf ${this.options.videoCrf ?? 23}`)

    if (outputFormat !== 'webm') {
      outputStream
        .outputOptions(`-preset ${this.options.videoPreset || 'ultrafast'}`)
        .outputOptions(`-pix_fmt ${this.options.videoPixelFormat || 'yuv420p'}`)
        .outputOptions(`-minrate ${this.options.videoBitrate || 1000}`)
        .outputOptions(`-maxrate ${this.options.videoBitrate || 1000}`)
        .outputOptions('-framerate 1')
    }

    if (outputFormat === 'webm') {
      outputStream
        .outputOptions('-quality realtime')
        .outputOptions('-speed 8')
        .outputOptions('-qmin 4')
        .outputOptions('-qmax 48')
        .outputOptions('-tile-columns 4')
        .outputOptions('-frame-parallel 1')
        .outputOptions('-row-mt 1')
        .outputOptions('-g 20')
    }

    outputStream.outputOptions(`-threads ${threads}`).on('progress', (progressDetails) => {
      this.duration = progressDetails.timemark
    })

    if (this.options.recordDurationLimit) {
      outputStream.duration(this.options.recordDurationLimit)
    }

    return outputStream
  }

  private get outputFormat(): SupportedVideoFileFormat {
    return this.options.outputFormat || 'mp4'
  }

  private handleWriteStreamError(errorMessage: string): void {
    this.emit('videoStreamWriterError', errorMessage)

    if (
      this.status !== 'inProgress' &&
      this.status !== 'stopping' &&
      errorMessage.includes('pipe:0: End of file')
    )
      return

    return console.error(`Error unable to capture video stream: ${errorMessage}`)
  }

  private findSlot(timestamp: number): number {
    if (this.screenCastFrames.length === 0) return 0

    let i: number
    let frame: PageScreenFrame

    for (i = this.screenCastFrames.length - 1; i >= 0; i--) {
      frame = this.screenCastFrames[i]

      if (timestamp > frame.timestamp) break
    }

    return i + 1
  }

  insert(frame: PageScreenFrame): void {
    if (this.status === 'stopping' || this.status === 'completed') return

    // reduce the queue into half when it is full
    if (this.screenCastFrames.length === this.screenLimit) {
      const numberOfFramesToSplice = Math.floor(this.screenLimit / 2)
      const framesToProcess = this.screenCastFrames.splice(0, numberOfFramesToSplice)
      this.processFrameBeforeWrite(framesToProcess, this.screenCastFrames[0].timestamp)
    }

    const insertionIndex = this.findSlot(frame.timestamp)

    if (insertionIndex === this.screenCastFrames.length) {
      this.screenCastFrames.push(frame)
    } else {
      this.screenCastFrames.splice(insertionIndex, 0, frame)
    }
  }

  async stop(stoppedTime = Date.now() / 1000): Promise<void> {
    if (!this.writerPromise)
      throw new Error('Writer promise not initialized: did startStreamWriter() throw?')

    if (this.status === 'completed' || this.status === 'stopping') return this.writerPromise

    this.status = 'stopping'
    this.drainFrames(stoppedTime)

    this.videoMediatorStream.end()
    this.status = 'completed'

    await this.writerPromise
  }

  private drainFrames(stoppedTime: number): void {
    this.processFrameBeforeWrite(this.screenCastFrames, stoppedTime)
    this.screenCastFrames = []
  }

  private processFrameBeforeWrite(frames: PageScreenFrame[], chunkEndTime: number): void {
    const processedFrames = this.trimFrame(frames, chunkEndTime)

    processedFrames.forEach(({ blob, duration }) => this.write(blob, duration))
  }

  private trimFrame(fameList: PageScreenFrame[], chunkEndTime: number): PageScreenFrame[] {
    return fameList.map((currentFrame: PageScreenFrame, index: number) => {
      const endTime = index !== fameList.length - 1 ? fameList[index + 1].timestamp : chunkEndTime
      const duration = endTime - currentFrame.timestamp

      return { ...currentFrame, duration }
    })
  }

  private write(data: Buffer, durationSeconds = 1): void {
    if (this.status === 'notStarted') this.status = 'inProgress'

    const numberOfFps = Math.max(Math.floor(durationSeconds * this.fps), 1)

    for (let i = 0; i < numberOfFps; i++) {
      if (this.videoMediatorStream.writable) this.videoMediatorStream.write(data)
    }
  }

  private async configureFfmpegPath(): Promise<void> {
    const ffmpegPath = await this.getFfmpegPath()

    if (!ffmpegPath) {
      throw new Error('FFmpeg path is missing, \n Set the FFMPEG_PATH env variable')
    }

    setFfmpegPath(ffmpegPath)
  }

  private async getFfmpegPath(): Promise<string | null> {
    if (this.options.ffmpegPath) return this.options.ffmpegPath

    try {
      // eslint-disable-next-line import/no-extraneous-dependencies
      const ffmpegInstaller = await import('@ffmpeg-installer/ffmpeg')
      if (ffmpegInstaller.path) return ffmpegInstaller.path

      return null
    } catch (e) {
      return null
    }
  }

  private get fps(): number {
    return this.options.fps ?? 25
  }

  private videoFileCodec(filePath: string): string {
    return this.options.videoCodec
      ? this.options.videoCodec
      : fileFormatFor(filePath) === 'webm'
      ? 'libvpx-vp9'
      : 'libx265'
  }

  private videoStreamCodec(): string {
    return this.options.videoCodec
      ? this.options.videoCodec
      : this.options.outputFormat === 'webm'
      ? 'libvpx-vp9'
      : 'libx265'
  }

  private get videoFrameSize(): string {
    const { width, height } = this.options.videoFrame || {}
    return typeof width === 'number' && typeof height === 'number' ? `${width}x${height}` : '100%'
  }

  private get autopad(): { activation: boolean; color?: string } {
    const autopad = this.options.autopad

    return !autopad ? { activation: false } : { activation: true, color: autopad.color }
  }
}

async function checkVideoCodec(codec: string) {
  const availableCodecsMap = await getAvailableCodecs()
  const found = availableCodecsMap[codec]

  if (!found || !found.canEncode) {
    const list = listAvailableEncodingCodecs(availableCodecsMap)
    throw new Error(`Video codec ${codec} is not available for encoding. Available codecs:\n${list}`)
  }
}

function getAvailableCodecs() {
  return new Promise<Codecs>((resolve, reject) =>
    ffmpeg().getAvailableCodecs((err, codecs) => {
      if (err) reject(err)
      else resolve(codecs)
    })
  )
}

function listAvailableEncodingCodecs(availableCodecsMap: Record<string, Codec>) {
  return Object.entries(availableCodecsMap)
    .filter(([, { canEncode }]) => canEncode)
    .map(([codec, { description }]) => `- ${codec}: ${description}`)
    .join('\n')
}
